### ISSUE TITLE
Release for Scala 2.13.0-M3, please?

### ISSUE THREAD

**Issue by SethTisue** [etorreborre/specs2#632]
if there are missing dependencies, let's get the ticket-opening, polite-nagging process started on those :-)

**Comment by etorreborre** [comment]
I indeed gave it a go 2 days ago, here is what's missing:

 [ ] shapeless
 [ ] kind-projector

Thanks for chasing those 2 dependencies, I still have very few open-source time for me these days.

**Comment by SethTisue** [comment]
Shapeless ticket: https://github.com/milessabin/shapeless/issues/811
kind-projector ticket: https://github.com/non/kind-projector/issues/63

**Comment by SethTisue** [comment]
Shapeless and kind-projector have now both been published for M3.

**Comment by etorreborre** [comment]
Thanks @SethTisue, I can indeed get those new dependencies. Now I have to fix a bunch of ambiguous implicits.

**Comment by etorreborre** [comment]
I'm getting a suspicious compiler error on this expression:
```scala
def step(a: =>Any): Unit =...
var stack: Stack = new Stack

def stackIsEmpty  = step(stack = emptyStack)
```
```
unknown parameter name: stack
```
But in that case `stack` is not a parameter name, it refers to a `var` in scope.

**Comment by SethTisue** [comment]
https://github.com/scala/scala/pull/6092 ? I think you'll need  to change it to curly braces `step{stack = emptyStack}`

**Comment by etorreborre** [comment]
Thanks, that's indeed what I did. I think it's a good idea to restrict this behaviour in 2.13. `4.0.2` is now available for scala `2.13.0-M3`: https://oss.sonatype.org/content/repositories/releases/org/specs2/specs2-core_2.13.0-M3/4.0.2

**Comment by etorreborre** [comment]
Hi @SethTisue there seems to be a change in the implicit resolution in 2.13. This test doesn't pass anymore: https://github.com/etorreborre/specs2/blob/efccac81c4784fb19cc081ff37fd2958f1c2f79d/tests/src/test/scala/org/specs2/matcher/describe/DiffableSpec.scala#L158. 

The compiler says that the implicit val is not used and indeed it should be used to make the test pass. Do you have any idea of what has changed here? Thanks

**Comment by SethTisue** [comment]
https://github.com/scala/scala/pull/6010? https://github.com/scala/scala/pull/6074? https://github.com/scala/scala/pull/5867? https://github.com/scala/scala/pull/6123? https://github.com/scala/scala/pull/6069?

(yes some of these are 2.12.x PRs, but they weren't enabled by default until M3)

**Comment by etorreborre** [comment]
I just did a bit of testing. I get the feeling that scala/scala#6010 might be involved but the behaviour is curious. For example the `Diffable` for
```
case class B(i: Int, as: List[A])
``` 
does not pick up the local implicit definition for `Diffable[A]` while this definition gets it:
```
case class B(as: List[A], i: Int)
```
This is for now beyond my implicit resolution powers :-). @milessabin do you have an idea of what could go wrong with the shapeless recursion here?

**Comment by milessabin** [comment]
Has this failure started happening _since_ 2.13.0-M3 was released? ie. a subsequent change is responsible?

**Comment by milessabin** [comment]
https://github.com/scala/scala/pull/6069 was included in 2.13.0-M3 and your earlier comment suggested your build was fine with that release ...

**Comment by etorreborre** [comment]
My build fails with `2.13.0-M3` because a local implicit is *not* picked up. It works fine with `2.12.4` so I am wondering if the fix for scala/scala#6010 could be a regression.

**Comment by milessabin** [comment]
Have you tried reprioritizing the instances? If this doesn't compile with 2.13.0-M3 then it looks like it might be more fallout from https://github.com/scala/scala/pull/6069 /cc @TomasMikula.

**Comment by TomasMikula** [comment]
Testing with 2.12.4 with the compiler option `-Xsource:2.13` could possibly narrow it down.

**Comment by hmemcpy** [comment]
For what it's worth, I managed to reproduce this, as @TomasMikula suggested, targeting 2.12.4 with `-Xsource:2.13` enabled. The issue is that the local implicit appears unused:
```
DiffableSpec.scala:158: local val diffableA in method ccFieldDiffable is never used
[error]     implicit val diffableA: Diffable[A] = new Diffable[A] {
[error]
```
Digging a bit deeper (and doing some println-debugging 😄) I was able to see that the resolution is identical between both versions:
```
CaseClassDiffable
 -> ClassFieldsDifferenceInspectable
   -> SeqLinesDiffable
      -> DiffableSpec$$anon$2@757e5d6f <-- the local implicit
```

Unfortunately, in the later case, the compiler reports that the local implicit instance is unused, but in actuality it is.

**Comment by hmemcpy** [comment]
Also confirm that when the order of parameters is reversed (via Eric's comment, `case class B(i: Int, as: List[A])` -> `case class B(as: List[A], i: Int)`), the warning does not occur, and the implicits graph looks like this:
```
CaseClassDiffable
 -> ClassFieldsDifferenceInspectable
   -> SeqLinesDiffable
      -> DiffableSpec$$anon$2@... <-- the local implicit
   -> PrimitiveDiffable
```

**Comment by etorreborre** [comment]
If the implicit is actually used and wrongly reported as unused how come the test doesn't pass in 2.13?

**Comment by hmemcpy** [comment]
It passes when I remove `-Xfatal-warnings` and run it. Here's what I'm doing (after removing the flag from build.sbt):
```
$ sbt
> ++ 2.13.0-M3
> project tests
> testOnly *DiffableSpec -- -ex "Allow using custom Diffable for case class fields"
... (lots of compiler output later)
[warn] .../DiffableSpec.scala:158: local val diffableA in method ccFieldDiffable is never used
[warn]     implicit val diffableA: Diffable[A] = new Diffable[A] {
[warn]                        ^
[warn] one warning found
...
[info]   ============
[info]
[info]   + Allow using custom Diffable for case class fields
[info]
[info] Total for specification DiffableSpec
[info] Finished in 330 ms
[info] 1 example, 0 failure, 0 error
[info]
[info] Passed: Total 1, Failed 0, Errors 0, Passed 1
```
Unless I'm holding it wrong, this is just a compiler warning :)
Removing the implicit value, of course, causes the test to fail for real.

**Comment by etorreborre** [comment]
@hmemcpy oh really? I didn't think of trying that, I just assumed that the unused warning was correct. That's a different story then. @SethTisue I should probably report this as a spurious warning but I'd need a reduced test case, right?

**Comment by SethTisue** [comment]
incorrect unused warnings aren't unheard of, the feature is relatively new and still being refined. perhaps if @som-snytt takes a look he can quickly intuit what's going on without you having to isolate it

**Comment by som-snytt** [comment]
Edit: I see you did a workaround on master. 

So you need `-Ywarn-macros:after` to look at trees after macro expansion. The default is to look at trees before macro expansion, for purposes of warnings. Before macro expansion, that value is unused.
```
$ scalac -Ywarn-macros:help
Usage: -Ywarn-macros:<mode> where <mode> choices are none, before, after, both (default: before).
  none    Do not inspect expansions or their original trees when generating unused symbol warnings.
  before  Only inspect unexpanded user-written code for unused symbols.
  after   Only inspect expanded trees when generating unused symbol warnings.
  both    Inspect both user-written code and expanded trees when generating unused symbol warnings.
```
Is this on master? I also couldn't reproduce with `2.12.4` and `-Xsource:2.13`. I do see the spurious warnings from `2.12.5` that are fixed in `2.12.6`. This is a fresh clone of specs2.

```
[info] Set current project to specs2 (in build file:/Users/andrew/workspace/specs2/)
specs2(master)> ++ 2.13.0-M3
[info] Setting version to 2.13.0-M3
[info] Reapplying settings...
[info] Set current project to specs2 (in build file:/Users/andrew/workspace/specs2/)
specs2(master)> project tests
[info] Set current project to specs2-tests (in build file:/Users/andrew/workspace/specs2/)
tests> testOnly *DiffableSpec -- -ex "Allow using custom Diffable for case class fields"
[snip]
[info]   
[info]   
[info]   todo: handle non explicit types for Failure !!!
[info] 
[info]   Case Classes
[info]   ============
[info] 
[info]   + Allow using custom Diffable for case class fields
[info] 
[info] Total for specification DiffableSpec
[info] Finished in 361 ms
[info] 1 example, 0 failure, 0 error
[info] 
[info] Passed: Total 1, Failed 0, Errors 0, Passed 1
[success] Total time: 151 s, completed Mar 29, 2018 11:40:28 AM
```
